operators/src/pro/cache/shared_cache.rs

```diff
@@ -0,0 +1,1190 @@
+            entry.query.is_match(query)
+        });
+
+        CacheQueryResult {
+            cache_hit: x.map(|(id, entry)| (*id, entry)),
+            expired_cache_entry_ids,
+        }
+    }
+}
+
+struct CacheQueryResult<'a, Query, CE> {
+    cache_hit: Option<(CacheEntryId, &'a CacheQueryEntry<Query, CE>)>,
+    expired_cache_entry_ids: Vec<CacheEntryId>,
+}
+
+pub trait Cache<C: CacheElement>:
+    CacheView<
+    CacheQueryEntry<C::Query, C::CacheContainer>,
+    CacheQueryEntry<C::Query, C::LandingZoneContainer>,
+>
+where
+    C::Query: Clone + CacheQueryMatch,
+    CacheQueryEntry<C::Query, C::LandingZoneContainer>: ByteSize,
+    CacheQueryEntry<C::Query, C::CacheContainer>: ByteSize,
+    CacheQueryEntry<C::Query, C::CacheContainer>:
+        From<CacheQueryEntry<C::Query, C::LandingZoneContainer>>,
+{
+    /// This method returns a mutable reference to the cache entry of an operator.
+    /// If there is no cache entry for the operator, this method returns None.
+    fn operator_cache_view_mut(
+        &mut self,
+        key: &CanonicOperatorName,
+    ) -> Option<OperatorCacheEntryView<C>>;
+
+    /// This method queries the cache for a given query.
+    /// If the query matches an entry in the cache, the cache entry is returned and it is promoted in the LRU.
+    /// If the query does not match an entry in the cache, None is returned. Also if a cache entry is found but it is expired, None is returned.
+    ///
+    /// # Errors
+    /// This method returns an error if the cache entry is not found.
+    ///
+    fn query_and_promote(
+        &mut self,
+        key: &CanonicOperatorName,
+        query: &C::Query,
+    ) -> Result<Option<C::ResultStream>, CacheError> {
+        let mut cache = self
+            .operator_cache_view_mut(key)
+            .ok_or(CacheError::OperatorCacheEntryNotFound)?;
+
+        let CacheQueryResult {
+            cache_hit,
+            expired_cache_entry_ids,
+        } = cache.find_matching_cache_entry_and_collect_expired_entries(query);
+
+        let res = if let Some((cache_entry_id, cache_entry)) = cache_hit {
+            let stream = cache_entry.elements.result_stream(query);
+
+            // promote the cache entry in the LRU
+            cache.lru.promote(&cache_entry_id);
+            Some(stream)
+        } else {
+            None
+        };
+
+        // discard expired cache entries
+        cache.discard_querys_from_cache_and_lru(&expired_cache_entry_ids);
+
+        Ok(res.flatten())
+    }
+
+    /// This method inserts a query into the cache.
+    ///
+    /// # Errors
+    /// This method returns an error if the query is already in the cache.
+    ///
+    fn insert_query_into_landing_zone(
+        &mut self,
+        key: &CanonicOperatorName,
+        query: &C::Query,
+    ) -> Result<QueryId, CacheError> {
+        self.create_operator_cache_if_needed(key.clone());
+        self.operator_cache_view_mut(key)
+            .expect("OperatorCache was Just created ")
```

Comment:
redundant space at the end and "Just" capitalized.
In general: we have a guideline for formulating expect-messages in the CODESTYLE.md, maybe try to stick to that more.

Reply:
b889d4d77a23979493248b5996133ca7fd520310

Reply:
[b889d4d](https://github.com/geo-engine/geoengine/pull/833/commits/b889d4d77a23979493248b5996133ca7fd520310)